File: apps/mobile-app/lib/utils/format.dart
import 'package:intl/intl.dart';

/// Amounts are stored in paise and always shown with Indian grouping.
String formatMoney(int amountMinor, [String currency = 'INR']) {
  final formatter = NumberFormat.currency(
    locale: 'en_IN',
    symbol: currency == 'INR' ? '₹' : '$currency ',
  );
  return formatter.format(amountMinor / 100);
}

/// Compact form for tiles and the month rollup: ₹42,318 with no paise.
String formatMoneyShort(int amountMinor, [String currency = 'INR']) {
  final formatter = NumberFormat.currency(
    locale: 'en_IN',
    symbol: currency == 'INR' ? '₹' : '$currency ',
    decimalDigits: 0,
  );
  return formatter.format(amountMinor / 100);
}

/// Everything is shown in IST, whatever the phone's own timezone says.
///
/// A ledger read on a phone that has picked up another country's clock —
/// or one simply set wrong — should still say a payment happened on the
/// evening of the 11th, because that is when it happened. India has a fixed
/// +05:30 offset and no daylight saving, so the arithmetic is safe.
const Duration _istOffset = Duration(hours: 5, minutes: 30);

/// The instant as IST wall-clock, carried in a DateTime whose own fields
/// are what should be displayed.
DateTime _ist(DateTime instant) => instant.toUtc().add(_istOffset);

/// The instant as IST wall-clock, for a date/time picker to edit. The
/// result is a local-flavoured DateTime whose fields read as IST.
DateTime istWallClock(DateTime instant) {
  final ist = _ist(instant);
  return DateTime(ist.year, ist.month, ist.day, ist.hour, ist.minute, ist.second);
}

/// The reverse: IST wall-clock fields back to the instant they name.
DateTime fromIstWallClock(DateTime wallClock) =>
    DateTime.utc(
      wallClock.year,
      wallClock.month,
      wallClock.day,
      wallClock.hour,
      wallClock.minute,
      wallClock.second,
    ).subtract(_istOffset);

/// Today's date in IST, as YYYY-MM-DD.
String istToday() => _ist(DateTime.now()).toIso8601String().substring(0, 10);

String formatDayLabel(String isoDate) {
  final today = istToday();
  if (isoDate == today) return 'Today';

  final yesterday = DateTime.parse('${today}T00:00:00Z')
      .subtract(const Duration(days: 1))
      .toIso8601String()
      .substring(0, 10);
  if (isoDate == yesterday) return 'Yesterday';

  // Parsed as a plain date, so no offset can shift which day is named.
  return DateFormat('EEE, d MMM').format(DateTime.parse(isoDate));
}

String formatTime(DateTime dateTime) => DateFormat('h:mm a').format(_ist(dateTime));

String formatDateTime(DateTime dateTime) => DateFormat('EEE, d MMM, h:mm a').format(_ist(dateTime));

/// "September 2026" for the analytics month picker.
String formatMonthLabel(String month) {
  final parts = month.split('-').map(int.parse).toList();
  return DateFormat('MMMM yyyy').format(DateTime(parts[0], parts[1]));
}

String shiftMonth(String month, int delta) {
  final parts = month.split('-').map(int.parse).toList();
  final shifted = DateTime(parts[0], parts[1] + delta);
  return '${shifted.year.toString().padLeft(4, '0')}-${shifted.month.toString().padLeft(2, '0')}';
}

String currentMonth() {
  // The IST month: at 1am on the 1st, UTC still says last month.
  return istToday().substring(0, 7);
}

File: apps/frontend/src/lib/format.ts
/** Amounts are stored in paise; always shown as ₹ with Indian grouping. */
export function formatMoney(amountMinor: number, currency = "INR"): string {
  return new Intl.NumberFormat("en-IN", { style: "currency", currency }).format(amountMinor / 100);
}

/** Compact form for tiles and rollups: ₹42,318 with no paise. */
export function formatMoneyShort(amountMinor: number, currency = "INR"): string {
  return new Intl.NumberFormat("en-IN", {
    style: "currency",
    currency,
    maximumFractionDigits: 0,
  }).format(amountMinor / 100);
}

/**
 * Everything is shown in IST, whatever the browser's own timezone is.
 *
 * A ledger read from a laptop abroad should still say a payment happened on
 * the evening of the 11th, because that is when it happened. Relying on the
 * device's clock would quietly renumber the days.
 */
const IST = "Asia/Kolkata";
const IST_OFFSET_MS = (5 * 60 + 30) * 60 * 1000;

/** Today's date in IST, as YYYY-MM-DD. */
export function istToday(): string {
  return new Date(Date.now() + IST_OFFSET_MS).toISOString().slice(0, 10);
}

export function formatDayLabel(isoDate: string): string {
  const today = istToday();
  if (isoDate === today) return "Today";

  const yesterday = new Date(Date.parse(`${today}T00:00:00Z`) - 24 * 60 * 60 * 1000)
    .toISOString()
    .slice(0, 10);
  if (isoDate === yesterday) return "Yesterday";

  // Read back at noon IST so the label can't slip a day either way.
  return new Date(`${isoDate}T12:00:00+05:30`).toLocaleDateString("en-IN", {
    weekday: "short",
    day: "numeric",
    month: "short",
    timeZone: IST,
  });
}

export function formatTime(iso: string): string {
  return new Date(iso).toLocaleTimeString("en-IN", {
    hour: "numeric",
    minute: "2-digit",
    timeZone: IST,
  });
}

export function formatDateTime(iso: string): string {
  return new Date(iso).toLocaleString("en-IN", {
    weekday: "short",
    day: "numeric",
    month: "short",
    hour: "numeric",
    minute: "2-digit",
    timeZone: IST,
  });
}

/** "September 2026" for the analytics month picker. */
export function formatMonthLabel(month: string): string {
  const [year, mon] = month.split("-").map(Number);
  return new Date(Date.UTC(year, mon - 1, 1)).toLocaleDateString("en-IN", {
    month: "long",
    year: "numeric",
    timeZone: "UTC",
  });
}

/** Shifts a YYYY-MM string by a number of months. */
export function shiftMonth(month: string, delta: number): string {
  const [year, mon] = month.split("-").map(Number);
  const date = new Date(Date.UTC(year, mon - 1 + delta, 1));
  return date.toISOString().slice(0, 7);
}

export function currentMonth(): string {
  // The IST month: at 1am on the 1st, UTC still says last month.
  return istToday().slice(0, 7);
}
